### ISSUE TITLE
base-model-interface-and-registry

### ISSUE THREAD

**Issue by louiswang524** [louiswang524/pytorch-recsys#3]
# Task: Base Model Interface and Registry System

## Objective
Implement the foundational model interface and registry system that all recommendation models will inherit from, ensuring consistent APIs and easy model discovery.

## Background
The base model interface defines the contract that all models in the framework must follow. The registry system provides automatic model discovery and instantiation capabilities, making it easy to switch between different model implementations through configuration.

### Base Model Interface
Create `pytorch_recsys/models/base.py` with:
- Abstract base class inheriting from PyTorch Lightning
- Standardized forward pass interface
- Prediction and inference methods
- Configuration validation
- Model serialization/deserialization

### Model Registry System
Create `pytorch_recsys/models/registry.py` with:
- Automatic model registration decorator
- Model discovery and instantiation
- Configuration-driven model creation
- Version and compatibility tracking

### Core Interfaces
All models must implement consistent interfaces for:
- Training and validation steps
- Prediction and recommendation generation  
- Model export for serving
- Hyperparameter management

### Base Model Class
```python
from abc import ABC, abstractmethod
from typing import Dict, Any, Optional, Tuple, List
import torch
from torch import nn, Tensor
import pytorch_lightning as pl
from omegaconf import DictConfig

class BaseSequentialModel(pl.LightningModule, ABC):
    """Abstract base class for sequential recommendation models"""
    
    def __init__(self, config: DictConfig):
        super().__init__()
        self.config = config
        self.save_hyperparameters()
        
        # Common parameters
        self.num_items = config.num_items
        self.num_users = config.get('num_users', None)
        self.max_seq_len = config.max_seq_len
        self.embedding_dim = config.embedding_dim
        
        # Validation
        self._validate_config()
        
        # Initialize common components
        self.item_embeddings = nn.Embedding(
            self.num_items + 1,  # +1 for padding token
            self.embedding_dim,
            padding_idx=0
        )
    
    @abstractmethod
    def forward(self, sequences: Tensor, **kwargs) -> Tensor:
        """Forward pass for the model"""
        pass
    
    @abstractmethod
    def predict_next_items(self, sequences: Tensor, k: int = 10) -> Tensor:
        """Generate top-k item predictions"""
        pass
    
    def training_step(self, batch: Dict[str, Tensor], batch_idx: int) -> Tensor:
        """Standard training step implementation"""
        sequences = batch['sequences']
        targets = batch['targets']
        
        logits = self(sequences)
        loss = self.compute_loss(logits, targets)
        
        self.log('train_loss', loss, on_step=True, on_epoch=True)
        return loss
    
    def validation_step(self, batch: Dict[str, Tensor], batch_idx: int) -> Tensor:
        """Standard validation step implementation"""
        sequences = batch['sequences']
        targets = batch['targets']
        
        logits = self(sequences)
        loss = self.compute_loss(logits, targets)
        
        self.log('val_loss', loss, on_step=False, on_epoch=True)
        return loss
    
    def configure_optimizers(self):
        """Configure optimizer and scheduler"""
        optimizer = torch.optim.Adam(
            self.parameters(),
            lr=self.config.learning_rate,
            weight_decay=self.config.get('weight_decay', 0.0)
        )
        
        scheduler_config = self.config.get('scheduler', None)
        if scheduler_config:
            scheduler = self._create_scheduler(optimizer, scheduler_config)
            return {
                'optimizer': optimizer,
                'lr_scheduler': {
                    'scheduler': scheduler,
                    'monitor': 'val_loss',
                    'interval': 'epoch',
                    'frequency': 1
                }
            }
        
        return optimizer
    
    @abstractmethod
    def compute_loss(self, logits: Tensor, targets: Tensor) -> Tensor:
        """Compute training loss"""
        pass
    
    def _validate_config(self):
        """Validate model configuration"""
        required_keys = ['num_items', 'max_seq_len', 'embedding_dim', 'learning_rate']
        for key in required_keys:
            if key not in self.config:
                raise ValueError(f"Missing required config key: {key}")
    
    def _create_scheduler(self, optimizer, scheduler_config):
        """Create learning rate scheduler"""
        scheduler_type = scheduler_config.type
        if scheduler_type == 'cosine':
            return torch.optim.lr_scheduler.CosineAnnealingLR(
                optimizer, T_max=scheduler_config.T_max
            )
        elif scheduler_type == 'step':
            return torch.optim.lr_scheduler.StepLR(
                optimizer, 
                step_size=scheduler_config.step_size,
                gamma=scheduler_config.gamma
            )
        else:
            raise ValueError(f"Unknown scheduler type: {scheduler_type}")
    
    def get_embedding_weights(self) -> Tensor:
        """Get item embedding weights for analysis"""
        return self.item_embeddings.weight.data
    
    def export_for_serving(self, export_path: str):
        """Export model for serving deployment"""
        # Will be implemented in serving module
        raise NotImplementedError("Export functionality not yet implemented")
```

### Model Registry
```python
from typing import Dict, Type, Any, Optional
from functools import wraps
import inspect
from omegaconf import DictConfig
from .base import BaseSequentialModel

class ModelRegistry:
    """Registry for automatic model discovery and instantiation"""
    
    _models: Dict[str, Type[BaseSequentialModel]] = {}
    _model_info: Dict[str, Dict[str, Any]] = {}
    
    @classmethod
    def register(cls, name: str, description: str = "", version: str = "1.0.0"):
        """Decorator for registering models"""
        def decorator(model_class: Type[BaseSequentialModel]):
            if not issubclass(model_class, BaseSequentialModel):
                raise ValueError(f"Model {model_class.__name__} must inherit from BaseSequentialModel")
            
            if name in cls._models:
                raise ValueError(f"Model {name} is already registered")
            
            cls._models[name] = model_class
            cls._model_info[name] = {
                'class': model_class,
                'description': description,
                'version': version,
                'module': model_class.__module__,
                'file': inspect.getfile(model_class)
            }
            
            return model_class
        return decorator
    
    @classmethod
    def create_model(cls, name: str, config: DictConfig) -> BaseSequentialModel:
        """Create model instance from configuration"""
        if name not in cls._models:
            available = list(cls._models.keys())
            raise ValueError(f"Model {name} not found. Available models: {available}")
        
        model_class = cls._models[name]
        return model_class(config)
    
    @classmethod
    def list_models(cls) -> Dict[str, Dict[str, Any]]:
        """List all registered models with their information"""
        return cls._model_info.copy()
    
    @classmethod
    def get_model_class(cls, name: str) -> Type[BaseSequentialModel]:
        """Get model class by name"""
        if name not in cls._models:
            raise ValueError(f"Model {name} not registered")
        return cls._models[name]
    
    @classmethod
    def is_registered(cls, name: str) -> bool:
        """Check if model is registered"""
        return name in cls._models

# Global registry instance
model_registry = ModelRegistry()

# Convenience decorator
def register_model(name: str, description: str = "", version: str = "1.0.0"):
    """Convenience decorator for model registration"""
    return model_registry.register(name, description, version)
```

### Configuration Validation
```python
from typing import Any, Dict, List
from omegaconf import DictConfig
from pydantic import BaseModel, validator

class ModelConfig(BaseModel):
    """Base configuration schema for all models"""
    
    # Required parameters
    num_items: int
    max_seq_len: int
    embedding_dim: int
    learning_rate: float
    
    # Optional parameters
    num_users: Optional[int] = None
    dropout_rate: float = 0.1
    weight_decay: float = 0.0
    batch_size: int = 256
    
    # Training parameters
    max_epochs: int = 100
    patience: int = 10
    gradient_clip_val: float = 1.0
    
    @validator('num_items')
    def validate_num_items(cls, v):
        if v <= 0:
            raise ValueError("num_items must be positive")
        return v
    
    @validator('max_seq_len')
    def validate_max_seq_len(cls, v):
        if v <= 0 or v > 10000:
            raise ValueError("max_seq_len must be between 1 and 10000")
        return v
    
    @validator('learning_rate')
    def validate_learning_rate(cls, v):
        if v <= 0 or v >= 1:
            raise ValueError("learning_rate must be between 0 and 1")
        return v

def validate_model_config(config: DictConfig) -> ModelConfig:
    """Validate and convert model configuration"""
    try:
        return ModelConfig(**config)
    except Exception as e:
        raise ValueError(f"Invalid model configuration: {e}")
```

### Functional Requirements
- [ ] BaseSequentialModel class implemented with all abstract methods
- [ ] ModelRegistry class with registration and instantiation capabilities
- [ ] Configuration validation with Pydantic schemas
- [ ] All common model operations (training_step, validation_step) implemented
- [ ] Proper error handling and validation

### Code Quality
- [ ] Full type hints for all methods and classes
- [ ] Comprehensive docstrings in Google format
- [ ] Abstract methods properly defined with ABC
- [ ] Consistent error messages and exception handling

### Testing
- [ ] Unit tests for BaseSequentialModel abstract interface
- [ ] Tests for ModelRegistry registration and instantiation
- [ ] Configuration validation tests
- [ ] Mock model implementation for testing

### Integration
- [ ] Integrates properly with PyTorch Lightning
- [ ] Compatible with Hydra configuration system
- [ ] Works with planned training and evaluation modules

### Unit Tests
```python
def test_base_model_interface():
    """Test abstract base model interface"""
    # Should not be instantiable
    with pytest.raises(TypeError):
        BaseSequentialModel(config)

def test_model_registry():
    """Test model registration and instantiation"""
    # Test registration
    # Test duplicate registration error
    # Test model creation
    # Test listing models

def test_config_validation():
    """Test configuration validation"""
    # Test valid configurations
    # Test invalid configurations
    # Test missing required fields
```

### Mock Implementation
Create a simple mock model for testing the interface:
```python
@register_model("mock_model", "Simple mock model for testing")
class MockSequentialModel(BaseSequentialModel):
    def forward(self, sequences: Tensor, **kwargs) -> Tensor:
        # Simple linear transformation for testing
        return self.item_embeddings(sequences).sum(dim=-2)
    
    def predict_next_items(self, sequences: Tensor, k: int = 10) -> Tensor:
        # Mock prediction implementation
        return torch.randint(1, self.num_items, (sequences.size(0), k))
    
    def compute_loss(self, logits: Tensor, targets: Tensor) -> Tensor:
        # Simple MSE loss for testing
        return nn.functional.mse_loss(logits, targets.float())
```

### Design Patterns
- **Template Method**: BaseSequentialModel defines the training flow
- **Registry Pattern**: Automatic model discovery and instantiation
- **Factory Pattern**: Model creation through configuration

### Performance Considerations
- Lazy loading of model classes
- Efficient configuration validation
- Minimal overhead in base class methods

### Extension Points
- Custom loss functions through method override
- Custom optimizers and schedulers
- Additional validation hooks

## Definition of Done
- All acceptance criteria met
- Unit tests passing with >95% coverage
- Mock model implementation works correctly
- Integration with PyTorch Lightning verified
- Documentation complete with usage examples
- Ready for actual model implementations (SASRec, TransAct)

This foundation enables rapid development of specific model implementations while maintaining consistency and quality across the framework.
